Editing "Update strategy" section (#2410)
elastic/cloud-on-k8s: docs/elasticsearch-specification.asciidoc

      maxSurge: 3
      maxUnavailable: 1
----

> `maxSurge`: Refers to the number of Pods that can be scheduled out of the total number of Pods in the currently applied specification. This setting is useful when the new specification has similar number of Pods, but different node sets. For example, when migrating Pods to a new node set. This setting prevents the operator from creating all new Pods before removing the old ones, by temporarily avoiding two times more Pods than the ones defined in the specification.

Reviewer: The original text was a bit confusing so the meaning got lost in the re-write. I would try to explain `MaxSurge` as follows.  
```suggestion
`maxSurge`: Refers to the number of extra Pods that can be temporarily scheduled exceeding the number of Pods  defined in the specification. This setting is useful for controlling the resource usage of the Kubernetes cluster when new Pods need to be spun up to replace existing Pods. For example, when applying a configuration change to a Node Set, ECK creates a clone of the Node Set with the desired configuration and shuts down the old Pods. Without the `MaxSurge` restriction, the cluster would temporarily contain twice as many Pods as usual while this operation is in progress.
```

`maxUnavailable`: Refers to the number of Pods that can be unavailable out of the total number of Pods in the currently applied specification. A Pod is defined unavailable when it is not ready from a Kubernetes perspective.

> As these constraints are enforced immediately after the specification is applied, it is possible that the actual state of the cluster might violate these constraints. When the cluster is outside these constraints, the operator can add or remove nodes only with the purpose of getting closer to these constraints. When the cluster is within these constraints, the operator can either add or remove nodes, as long as the cluster remains within the predefined constraints.

Reviewer: ```suggestion
The operator only tries to apply these constraints when a new specification is being applied. It is possible that the cluster state does not conform to the constraints at the beginning of the operation due to external factors. The operator will attempt to get to the desired state by adding or removing Pods as necessary while ensuring that the constraints are still satisfied.
```

      maxUnavailable: 1
----

> `maxSurge` is unbounded: This means that all the missing Pods are immediately created according to the best practices at any given time.

Reviewer: ```suggestion
`maxSurge` is unbounded: This means that all the required Pods are created immediately.
```

> `maxUnavailable` defaults to `1`: This assures that clusters have at most one unavailable Pod at any given time.

Reviewer: ```suggestion
`maxUnavailable` defaults to `1`: This ensures that the cluster has no more than one unavailable Pod at any given point in time.
```

==== Caveats
* With both `maxSurge` and `maxUnavailable` set to `0`, the operator cannot bring down an existing Pod nor create a new Pod.

> * Due to the safety measures employed by the operator, a particular `changeBudget` might be insufficient to perform the requested update. For example, with `maxSurge` set to 0 you cannot remove the last data node from one `nodeSet` and add a data node to a different `nodeSet`. In this case, the operator cannot remove the old node and migrate data as other nodes are not available, and can't add the new node because `maxSurge` is set to 0.

Reviewer: ```suggestion
* Due to the safety measures employed by the operator, certain `changeBudget`s might prevent the operator from making any progress . For example, with `maxSurge` set to 0, you cannot remove the last data node from one `nodeSet` and add a data node to a different `nodeSet`. In this case, the operator cannot create the new node because `maxSurge` is 0, and it cannot remove the old node because there are no other data nodes to migrate the data to.
```

> * Currently, the operator has a limited understanding of the cluster topology and during the update it might not pick the optimal order of Pods being added or removed. This can block the update process and you need to specify a higher `maxSurge` than theoretically needed.

Reviewer: ```suggestion
* For certain complex configurations, the operator might not be able to deduce the optimal order of operations necessary to achieve the desired outcome. If progress is blocked,  you may need to update the `maxSurge` setting to a higher value than the theoretical best to help the operator make progress in that case.
```
